src/dividat-driver/flex/main.go

```diff
@@ -0,0 +1,262 @@
+// package units into a buffer.
+func connectSerial(ctx context.Context, logger *logrus.Entry, serialName string, onReceive func([]byte)) {
+	mode := &serial.Mode{
+		BaudRate: 115200,
+		Parity:   serial.NoParity,
+		DataBits: 8,
+		StopBits: serial.OneStopBit,
+	}
+
+	START_MEASUREMENT_CMD := []byte{'S', '\n'}
+
+	logger.WithField("name", serialName).Info("Attempting to connect with serial port.")
+	port, err := serial.Open(serialName, mode)
+	if err != nil {
+		logger.WithField("config", mode).WithField("error", err).Info("Failed to open connection to serial port.")
+		return
+	}
+	defer port.Close()
+
+	_, err = port.Write(START_MEASUREMENT_CMD)
+	if err != nil {
+		logger.WithField("error", err).Info("Failed to write start message to serial port.")
+		port.Close()
+		return
+	}
+
+	reader := bufio.NewReader(port)
+	state := WAITING_FOR_HEADER
+	pointsLeftInSet := 0
+	bytesLeftInPoint := 0
+
+	var buff []byte
+	for {
+		// Terminate if we were cancelled
+		if ctx.Err() != nil {
+			logger.WithField("name", serialName).Info("Disconnecting from serial port.")
+			return
+		}
+
+		input, err := reader.ReadByte()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			continue
+		}
+
+		// Finite State Machine for parsing byte stream
+		switch {
+		case state == WAITING_FOR_HEADER && input == HEADER_START_MARKER:
+			state = HEADER_START
+		case state == HEADER_START && input == '\n':
+			state = HEADER_READ_LENGTH_MSB
+                case state == HEADER_READ_LENGTH_MSB:
+			// The number of measurements in each set may vary and is
+			// given as two consecutive bytes (big-endian).
+			msb := input
+			lsb, err := reader.ReadByte()
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				continue
+			}
+			pointsLeftInSet = int(binary.BigEndian.Uint16([]byte{msb,lsb}))
+			state = WAITING_FOR_BODY
+		case state == WAITING_FOR_BODY && input == BODY_START_MARKER:
+			state = BODY_START
+		case state == BODY_START && input == '\n':
+			state = BODY_POINT
+			bytesLeftInPoint = 4
+                case state == BODY_POINT:
+			buff = append(buff, input)
+			bytesLeftInPoint = bytesLeftInPoint - 1
+
+			if bytesLeftInPoint <= 0 {
+				pointsLeftInSet = pointsLeftInSet - 1
+
+				if pointsLeftInSet <= 0 {
+					// Finish and send set
+					onReceive(buff)
+
+					// Get ready for next set and request it
+					buff = []byte{}
+					state = WAITING_FOR_HEADER
+					_, err = port.Write(START_MEASUREMENT_CMD)
+					if err != nil {
+						logger.WithField("error", err).Info("Failed to write poll message to serial port.")
+						port.Close()
+						return
+					}
+				} else {
+					// Start next point
+					bytesLeftInPoint = 4
+				}
+			}
+		case state == UNEXPECTED_BYTE && input == HEADER_START_MARKER:
+			// Recover from error state when a new header is seen
+			buff = []byte{}
+			bytesLeftInPoint = 0
+			pointsLeftInSet = 0
+			state = HEADER_START
+		default:
+			state = UNEXPECTED_BYTE
```

Comment:
Debug log unexpected bytes ? Or would it be too chatty ?

Reply:
I don't see the need at this moment. If there was a serious issue there, we would probably further instrument the driver to observe.